### ISSUE TITLE
readonly property must be assigned in constructor

### ISSUE THREAD

**Issue by dungdm93** [microsoft/TypeScript#11579]
**TypeScript Version:**  2.0.3

**Code**

``` ts
class Person {
    public readonly name: string;
}
let p: Person = new Person();
// p.name ????
```

**Expected behavior:**
Compilation error. `Person.name` is not initialized.
So the `Person` class should be look like:

``` ts
class Person {
    public readonly name: string;
    constructor(name: string) {
        this.name = name;
    }
}

let p: Person = new Person("Superman");
```

-- or --

``` ts
class Person {
    public readonly name: string = "Superman";
}
```

**Actual behavior:** 
Compilation success. No errors/warnings

**Comment by jesseschalken** [comment]
See #8476

**Comment by dungdm93** [comment]
I think `readonly` is stricter. And should emit error even without `--strictNullChecks` flag

**Comment by jesseschalken** [comment]
> I think `readonly` is stricter.

Stricter in how the property can be used. That a property should be initialised so that when reading it you get a value of the specified type applies to both `readonly` and writable properties.

> And should emit error even without `--strictNullChecks` flag, isn't it?

If a property isn't initialised it will produce `undefined` when read. Without `--strictNullChecks`, all types implicitly include `undefined` (and `null`), so an uninitialised property always satisfies it's type annotation. Whether the property is `readonly` or not is irrelevant.

**Comment by DanielRosenwasser** [comment]
This seems reasonable. Basically the rule would be for a given property:

1. If the property is marked `readonly`
1. if the property has no initializer and
1. If the property has no type annotation, or if the given type is not undefined, and
1. if the property's containing class is non-ambient and does not have a constructor

then the compiler should issue an error. I think this could be a good first PR.

**Comment by jesseschalken** [comment]
> and does not have a constructor

So just adding `constructor () {}` would fix the compiler error but still leave the code broken?

**Comment by DanielRosenwasser** [comment]
Good catch 😃

**Comment by jarrydfillmore** [comment]
cool!

**Comment by masaeedu** [comment]
@DanielRosenwasser Would @helmutschneider's change in https://github.com/helmutschneider/TypeScript/commit/af9e5fbae work here? It could be modified to apply only to `readonly`, but this leaves non-readonly properties as a source of `undefined` contamination, even with `--strictNullChecks` on.

**Comment by superamadeus** [comment]
Seems like this is fixed by strict property initialization
